core-v1/src/main/scala/data/Justified.scala

```diff
@@ -293,6 +293,46 @@ object Justified {
   implicit def wrapQuantifier[OP[_]]: WrapQuantifier[Justified, OP] =
     WrapQuantifierJustified.asInstanceOf[WrapQuantifier[Justified, OP]]
 
+  implicit val semigroupal: Semigroupal[Justified] = new Semigroupal[Justified] {
+    override def product[A, B](
+      fa: Justified[A],
+      fb: Justified[B],
+    ): Justified[(A, B)] = {
+      Justified.byInference("product", (fa.value, fb.value), NonEmptySeq(fa, Vector(fb)))
+    }
+  }
+
+  implicit val traverse: Traverse[Justified] = new Traverse[Justified] {
+
+    override def map[A, B](fa: Justified[A])(f: A => B): Justified[B] = {
+      val derived = f(fa.value)
+      Justified.byInference("map", derived, NonEmptySeq.of(fa))
+    }
+
+    override def traverse[G[_] : Applicative, A, B](fa: Justified[A])(f: A => G[B]): G[Justified[B]] = {
+      val gb = f(fa.value)
+      Applicative[G].map(gb) { b =>
+        Justified.byInference("traverse", b, NonEmptySeq.of(fa))
+      }
+    }
```

Comment:
These are very generic names for justification... I wish I could avoid them to be more specific, but the recursive nature of the type definition of `HList` and thus `ExprHList` means that I can only build this chain of justification one node at a time, so we end up with a tree of "map" inferences for each `Expr` node of the `ExprHList`. The best I can do is avoid using this `Functor` / `Traverse` instance for anything other than the building of `HList` outputs and then I can rename `"map"` to something more specific to `HList` like `"::"` or `"cons"` or `"concat"`. For now, I'm gonna keep the generic definition and work backwards to replace anything it is used for later.